util/caller/resolver.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package caller
+
+import (
+	"os"
+	"regexp"
+	"runtime"
+	"strings"
+	"sync"
+)
+
+type cachedLookup struct {
+	file string
+	line int
+	fun  string
+}
+
+var dummyLookup = cachedLookup{file: "???", line: 1, fun: "???"}
+
+// A CallResolver is a helping hand around runtime.Caller() to look up file,
+// line and name of the calling function. CallResolver caches the results of
+// its lookups and strips the uninteresting prefix from both the caller's
+// location and name; see NewCallResolver().
+type CallResolver struct {
+	mu    sync.Mutex
+	cache map[uintptr]*cachedLookup
+	re    *regexp.Regexp
+}
+
+// defaultPattern strips src/github.com/organization/project/module/submodule/file.go
+// to module/submodule/file.go.
+var defaultPattern = "src/" + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, string(os.PathSeparator)), 3) + "(.*)"
```

Comment:
@tamird: It looks like `runtime.Caller` already strips `$GOPATH`; we get paths starting with `src/`. Good point about clickability; we could use a delimiter other than `/` if we still want to go down that route.

@tschottdorf I'd like to make truncation a display-time filter so that things like our web-based log viewer can easily generate links to the source of a file; the same argument applies to tracers and tasks. (And I think logs may be the only case where horizontal space is at a premium, so we may not even need to truncate in those cases)